### PR TITLE
Add a check for not templateable fields

### LINKED ISSUE #29819: DAG fails serialization if template_field contains execution_timeout
### Apache Airflow version

2.5.1

### What happened

If an Operator specifies a template_field with `execution_timeout` then the DAG will serialize correctly but throw an error during deserialization. This causes the entire scheduler to crash and breaks the application.

### What you think should happen instead

The scheduler should never go down because of some code someone wrote, this should probably throw an error during serialization.

### How to reproduce

Define an operator like this
```
class ExecutionTimeoutOperator(BaseOperator):
    template_fields = ("execution_timeout", )

    def __init__(self, execution_timeout: timedelta, **kwargs):
        super().__init__(**kwargs)
        self.execution_timeout = execution_timeout
```
then make a dag like this
```
dag = DAG(
    "serialize_with_default",
    schedule_interval="0 12 * * *",
    start_date=datetime(2023, 2, 28),
    catchup=False,
    default_args={
        "execution_timeout": timedelta(days=4),
    },
)

with dag:
    execution = ExecutionTimeoutOperator(task_id="execution", execution_timeout=timedelta(hours=1))
```
that will break the scheduler, you can force the stack trace by doing this
```
from airflow.models import DagBag
db = DagBag('dags/', read_dags_from_db=True)
db.get_dag('serialize_with_default')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/dagbag.py", line 190, in get_dag
    self._add_dag_from_db(dag_id=dag_id, session=session)
  File "/usr/local/lib/python3.9/site-packages/airflow/models/dagbag.py", line 265, in _add_dag_from_db
    dag = row.dag
  File "/usr/local/lib/python3.9/site-packages/airflow/models/serialized_dag.py", line 218, in dag
    dag = SerializedDAG.from_dict(self.data)
  File "/usr/local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 1287, in from_dict
    return cls.deserialize_dag(serialized_obj["dag"])
  File "/usr/local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 1194, in deserialize_dag
    v = {task["task_id"]: SerializedBaseOperator.deserialize_operator(task) for task in v}
  File "/usr/local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 1194, in <dictcomp>
    v = {task["task_id"]: SerializedBaseOperator.deserialize_operator(task) for task in v}
  File "/usr/local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 955, in deserialize_operator
    cls.populate_operator(op, encoded_op)
  File "/usr/local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 864, in populate_operator
    v = cls._deserialize_timedelta(v)
  File "/usr/local/lib/python3.9/site-packages/airflow/serialization/serialized_objects.py", line 513, in _deserialize_timedelta
    return datetime.timedelta(seconds=seconds)
TypeError: unsupported type for timedelta seconds component: str
```

### Operating System

Mac 13.1 (22C65)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==5.1.0
apache-airflow-providers-apache-hdfs==3.2.0
apache-airflow-providers-apache-hive==5.1.1
apache-airflow-providers-apache-spark==4.0.0
apache-airflow-providers-celery==3.1.0
apache-airflow-providers-cncf-kubernetes==5.1.1
apache-airflow-providers-common-sql==1.3.3
apache-airflow-providers-datadog==3.1.0
apache-airflow-providers-ftp==3.3.0
apache-airflow-providers-http==4.1.1
apache-airflow-providers-imap==3.1.1
apache-airflow-providers-jdbc==3.3.0
apache-airflow-providers-jenkins==3.2.0
apache-airflow-providers-mysql==4.0.0
apache-airflow-providers-pagerduty==3.1.0
apache-airflow-providers-postgres==5.4.0
apache-airflow-providers-presto==4.2.1
apache-airflow-providers-slack==7.2.0
apache-airflow-providers-sqlite==3.3.1
apache-airflow-providers-ssh==3.4.0

### Deployment

Docker-Compose

### Deployment details

I could repro this with docker-compose and in a helm backed deployment so I don't think it's really related to the deployment details

### Anything else

In the serialization code there are two pieces of logic that are in
direct conflict with each other. The first dictates how template fields
are serialized, from the code

```
# Store all template_fields as they are if there are JSON Serializable
# If not, store them as strings
```

and the second special cases a few names of arguments that need to be
deserialized in a specific way

```
elif k in {"retry_delay", "execution_timeout", "sla", "max_retry_delay"}:
    v = cls._deserialize_timedelta(v)
```

so during serialization airflow sees that execution_timeout is a
template field, serializes it as a string, then during deserialization
it is a special name that forces the deserialization as timedelta and BOOM!

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by hussein-awala**
closes: #29819

---
There are some fields processed by the scheduler and they are not templateable, but currently we don't have any check for these fields. In this PR extract the parameters list for the method `__init__` of `BaseOperator` and I check if there is a field in the `template_fields` in this list to raise an exception, where all the `BaseOperator` parameters are processed and used by the scheduler during processing time.

```python
>>> import inspect
>>> from airflow.models import BaseOperator
>>> set(inspect.signature(BaseOperator.__init__).parameters.keys())
{'start_date', 'doc', 'executor_config', 'pre_execute', 'email', 'retry_exponential_backoff', 'depends_on_past', 'inlets', 'wait_for_downstream', 'email_on_failure', 'outlets', 'doc_json', 'execution_timeout', 'task_id', 'sla', 'on_success_callback', 'trigger_rule', 'ignore_first_depends_on_past', 'doc_md', 'kwargs', 'pool', 'doc_rst', 'on_failure_callback', 'email_on_retry', 'weight_rule', 'self', 'priority_weight', 'pool_slots', 'run_as_user', 'task_group', 'doc_yaml', 'on_retry_callback', 'do_xcom_push', 'max_retry_delay', 'retries', 'owner', 'params', 'default_args', 'queue', 'resources', 'task_concurrency', 'max_active_tis_per_dag', 'wait_for_past_depends_before_skipping', 'retry_delay', 'on_execute_callback', 'dag', 'end_date', 'post_execute'}
```

### CODE REVIEW

File: airflow/serialization/serialized_objects.py [hunk 1]
```diff
@@ -770,8 +772,12 @@ def _serialize_node(cls, op: BaseOperator | MappedOperator, include_deps: bool)
 
         # Store all template_fields as they are if there are JSON Serializable
         # If not, store them as strings
+        # And raise an exception if the field is not templateable
+        forbidden_fields = set(inspect.signature(BaseOperator.__init__).parameters.keys())
         if op.template_fields:
             for template_field in op.template_fields:
+                if template_field in forbidden_fields:
+                    raise AirflowException(f"Cannot use {template_field} as template field")
```
Comment (eladkal) [inline airflow/serialization/serialized_objects.py]: I think the only case where we can not template fields is when user wants to template BaseOperator parameters which is something that can't be done. So maybe we can be more informative on the exception here?

Comment (hussein-awala) [reply airflow/serialization/serialized_objects.py]: I updated the exception message to `Cannot template BaseOperator fields: {template_field}`, do you have another suggestion?

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Nice one!

**Comment by Inetov** [comment]
@hussein-awala what if my custom operator needs to use templating for "params" field?
how to bypass the restriction introduced in this commit?

**Comment by hussein-awala** [comment]
> @hussein-awala what if my custom operator needs to use templating for params field?
how to bypass the restriction introduced in this commit?

@Inetov This is not possible for now, I will check if templating `params` (and other parameters) is 100% safe, and if it's the case, maybe we can remove them from the excluded fields list.

**Comment by eladkal** [comment]
@Inetov @hussein-awala see feature request for this https://github.com/apache/airflow/issues/31801

**Comment by Inetov** [comment]
thanks @eladkal, but #31801 is still in the process of being solved, and I can no longer upgrade to 2.5.2 and higher.
@hussein-awala, maybe there is a temporary workaround?

**Comment by potiuk** [comment]
I think that's a good idea to make a PR to exclude some of the fields in this way. I think PR for that would be good @gdavoian

**Comment by gdavoian** [comment]
@potiuk do you have any ideas on how it would fit into `BaseOperator`? Because I can only think about something like `templateable_fields` (the exact naming isn't important at the moment) to exclude them from the check. But simultaneously having two entities like `template_fields` and `templateable_fields` would be very misleading: the former is a list of fields that **can be** templated and the latter is a list of fields that **are allowed to be** templated, meaning that some fields (e.g., `BaseOperator.email`) have to be added to both lists at the same time... Again, the exact naming doesn't matter here as soon as the whole idea is weird.

And that actually leads to my initial question: why did we ever need that check in the first place? `BaseOperator` doesn't define any `template_fields` by default, leaving that up to developers of third-party/custom operators to do that, which is completely fine. Thus, it's the responsibility of those developers to ensure their operators are valid and serializable, isn't it?

**Comment by hussein-awala** [comment]
@gdavoian I agree with Jarek.

This PR was created to avoid the rendering issues with BaseOperator attributes like `execution_timeout`.

IMHO, you can create a test that loops over the excluded attributes, try to use a templated value, and check the value in runtime after rendering it. To know what attributes could be excluded, you can start by excluding all the params and keeping only the params that pass the test.

**Comment by gdavoian** [comment]
> This PR was created to avoid the rendering issues with BaseOperator attributes like execution_timeout.

But how does this PR help solve the rendering issue? I understand that it tries to warn the users against their potentially mistaken intentions, though in fact it simply replaces one type of exception raised with another, reduces flexibility, and introduces a backward-incompatible change as a side effect. So the person who tried to add `execution_timeout` to `template_fields` still won't be able to do that, whereas those who are using Jinja templates for rendering some of the `BaseOperator` fields (e.g., `email`, `pool`, `doc_md`) in their custom operators simply won't be able to upgrade their Airflow environments since their working code will suddenly turn out to be broken. Thus, honestly, I don't see any benefits from this change at all...

> IMHO, you can create a test that loops over the excluded attributes, try to use a templated value, and check the value in runtime after rendering it. To know what attributes could be excluded, you can start by excluding all the params and keeping only the params that pass the test.

This sounds like a workaround for a workaround :) For some reason, you've prohibited the rendering of all the `BaseOperator` fields (I'm still not convinced of the need for this change), and now you're asking me to exclude some of the fields from that rule (the rule I don't even agree and the rule you can't really justify)...

**Comment by potiuk** [comment]
@gdavoidan -  what I proposed and I believe @hussein-awala agrees with is that you define which fields of BaseOperator you exclude from the check. This is what my proposal is about. Find out which of those fields CAN be templated and remove them from the check. Basically allow-list of fields you want to allow templating. Rather than allow all by default. There is no need "specify" them as extra field - just hard-code them.

What @hussein-awala proposes is that you attempt to test it and figure out which fields can be templated by templating then one by one and trying to serialize the DAG. because not all of them CAN be serialized and you need to serialize DAG to DB to execute it. See description ogf the original error https://github.com/apache/airflow/issues/29819 whcih explains the reason for doing it. Try to do what the author of the oriignal issue and see it for yourself when you remoge the protection. You don't need to believe anyone's word just try it and read the issues.

> This sounds like a workaround for a workaround :) For some reason, you've prohibited the rendering of all the BaseOperator fields (I'm still not convinced of the need for this change), and now you're asking me to exclude some of the fields from that rule (the rule I don't even agree and the rule you can't really justify)...

No. You simply didn't dig deep enough. Look at the issue and try to understand what happened there. This is solution to a real problem that overcautiously excluded all the fields. What we are kindly asking you to do is to find out which fields out of those can be templated.

Just looking at your examples neither `pool` nor `doc_md` should be allowed to be tempalted. The`pool` field is evaluated by the scheduler when scheduler scheduler tasks for execution, and then when celery worker picks up the task it only selects the tasks belonging to it's queue. And it happens long before template rendering happens - template rendering happpen in the worker of cellery AFTER the worker already used `queue` field to pick tasks. Similary `doc_md` should not be templated, because it is used by the UI to display task details NOT task instance details - and in order to do do it, cannot render the field, because  again - rendering happens at execution time by the task, not when the UI decides to display description of the task. Similarly `execution_timeout` cannot be rendered, because it is serialized to the DB by DAG file processor, WAY BEFORE the task gets executed, and there it is stored as INT not string (which is the original issue that this PR solved). And it makes completely no sense to render `execution_timeout` because again, it is used BEFORE the task starts and rendering is happening in the worker - because execution_timout is applied at the monitoring level of the task in teh process that starts the actual "local task process" where rendering happen. So basically (if you look at the code) SIGARM is started and alarm on timeout is set before JINJA rendering happens. You likely would not like to do:

```
signal.setitimer(signal.ITIMER_REAL, '{{ renerd_timer }}')
```

That simply would not work.

So assumption when that PR got merged that most (if not all)  of the fields of BaseOperators are like that and all of them should be excluded by default. 

You seem to find one that might be excluded. Maybe more can be as well. What we are kindly asking you to review all of those, possibly test (as @hussein-awala proposed) and come with PR where you specifically exclude those fields that CAN be allowlisted and you tested that they work.

Is it too much of an ask?

**Comment by hussein-awala** [comment]
> But how does this PR help solve the rendering issue?

Most of these attributes should not be used as a templated field, so the PR helped quickly fail the dag parsing instead of falling in runtime.

> I'm still not convinced of the need for this change

Most of the BaseOperator fields are used in parsing time and not runtime, for example, execution_date, priority_weight, the different callbacks, the different dependencies params, retry params, and the pool (I don't know why you think it could be templated, it's used to see if we can queue the task or not, so before executing the job), and even doc_md which is a tasks param more than task instance param.

For the email, we used it in the TaskInstance class when the TI fails, so after executing it, it's ok to exclude it. I'm not sure if we can exclude the other params; as I mentioned, most of them are used by the scheduler, executor, or webserver before executing the task.

**Comment by gdavoian** [comment]
@potiuk @hussein-awala Thank you for the explanations. I'm starting to understand the problem and what I need to do. I'll check and get back to you soon. My apologies for any earlier misunderstandings.

**Comment by potiuk** [comment]
> @potiuk @hussein-awala Thank you for the explanations! I'm starting to understand the problem and what I'm supposed to do. I'll check and get back to you soon. My apologies for any earlier misunderstandings!

No worries :). Airflow is pretty ... complicated. ... and even after years of working on it every day I find something new or make wrong assumption. See that one for example https://github.com/apache/airflow/pull/35539#issuecomment-1802996355  just about 2 minutes ago. ...

**Comment by gdavoian** [comment]
Hm, you were right, looks like only `email` is a field that is 100% safe to template (as I mentioned before, the field is heavily used by our team as `default_args["email"]` for each and every DAG). There is also `run_as_user` that seems to be safe as well, but I've never used it, so can't really comment on its usage. If you're fine with that, then I can open a PR to exclude `email` (and optionally `run_as_user` if you think it makes sense) from that check.

**Comment by gdavoian** [comment]
@potiuk @hussein-awala You're welcome to review https://github.com/apache/airflow/pull/35546 :)

**Pull request merged**